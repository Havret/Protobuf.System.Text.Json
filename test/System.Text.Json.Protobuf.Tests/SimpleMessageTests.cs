using Shouldly;
using SmartAnalyzers.ApprovalTestsExtensions;
using Xunit;

namespace System.Text.Json.Protobuf.Tests;

public class SimpleMessageTests
{
    [Fact]
    public void Should_serialize_message_with_primitive_types()
    {
        // Arrange
        var msg = new SimpleMessage
        {
            Int32Property = 1,
            Int64Property = 2,
        };
        var jsonSerializerOptions = CreateJsonSerializerOptions();

        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }

    [Fact]
    public void Should_deserialize_message_with_primitive_types()
    {
        // Arrange
        var msg = new SimpleMessage
        {
            Int32Property = 1,
            Int64Property = 2,
        };
        var jsonSerializerOptions = CreateJsonSerializerOptions();

        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<SimpleMessage>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.Int32Property.ShouldBe(msg.Int32Property);
        deserialized.Int64Property.ShouldBe(msg.Int64Property);
    }

    [Fact]
    public void Should_serialize_message_with_repeated_list()
    {
        var msg = new MessageWithRepeatedList
        {
            Int32List = {1, 2, 3}
        };
        
        var jsonSerializerOptions = CreateJsonSerializerOptions();

        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);

        // Assert
        var approver = new ExplicitApprover();
        approver.VerifyJson(serialized);
    }
    
    [Fact]
    public void Should_deserialize_message_with_repeated_list()
    {
        var msg = new MessageWithRepeatedList
        {
            Int32List = {1, 2, 3}
        };
        
        var jsonSerializerOptions = CreateJsonSerializerOptions();

        var serialized = JsonSerializer.Serialize(msg, jsonSerializerOptions);
        var deserialized = JsonSerializer.Deserialize<MessageWithRepeatedList>(serialized, jsonSerializerOptions);

        // Assert
        deserialized.ShouldNotBeNull();
        deserialized.Int32List.ShouldBeEquivalentTo(msg.Int32List);
    }

    private static JsonSerializerOptions CreateJsonSerializerOptions()
    {
        var jsonSerializerOptions = new JsonSerializerOptions();
        jsonSerializerOptions.Converters.Add(new ProtobufJsonConverterFactory());
        return jsonSerializerOptions;
    }
}