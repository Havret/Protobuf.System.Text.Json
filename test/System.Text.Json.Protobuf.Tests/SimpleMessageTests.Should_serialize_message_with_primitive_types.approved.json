{
  "doubleProperty": 2.5,
  "floatProperty": 3.5,
  "int32Property": 1,
  "int64Property": 2,
  "uint32Property": 3,
  "uint64Property": 4,
  "sint32Property": 5,
  "sint64Property": 6,
  "fixed32Property": 7,
  "fixed64Property": 8,
  "sfixed32Property": 9,
  "sfixed64Property": 10,
  "boolProperty": true,
  "stringProperty": "hello"
}